File: 234_palindrome_linked_list/isPalindrome.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
from typing import Optional


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        slow = fast = head

        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next

        tail = self.reverse(slow)

        while head and tail:
            if head.val != tail.val:
                return False
            head = head.next
            tail = tail.next
        return True
        
    def reverse(self, head: Optional[ListNode]) ->  Optional[ListNode]:
        prev = None
        curr = head

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp

        return prev

File: 234_palindrome_linked_list/is_palindrome.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn is_palindrome(mut head: Option<Box<ListNode>>) -> bool {
        let mut len = 0;
        let mut cur = &head;
        while let Some(node) = cur {
            cur = &node.next;
            len += 1;
        }

        let mut cur = &mut head;
        for _ in 0..(len - 1) / 2 {
            cur = &mut cur.as_mut().unwrap().next;
        }

        let mut prev = None;
        let mut cur = cur.as_mut().unwrap().next.take();
        while let Some(mut node) = cur {
            cur = node.next;
            node.next = prev;
            prev = Some(node);
        }

        let mut right_part = &prev;
        let mut left_part = &head;
        while let (Some(r), Some(l)) = (right_part, left_part) {
            if r.val != l.val {
                return false;
            }
            right_part = &r.next;
            left_part = &l.next;
        }
        true
    }
}